### PR TITLE
[Dev] JIT for MoE router and preprocess

### PR DESCRIPTION
**Pull request by yaox12**
PR to the main branch: https://github.com/NVIDIA/Megatron-LM/pull/1919

- Enable JIT (`torch.compile`) for expert bias and MoE preprocess.
- Let force load balancing compatible with CUDA Graph.

### CODE REVIEW

**Comment by hxbai** [comment]
LGTM.

File: megatron/core/transformer/moe/moe_utils.py [hunk 1]
```diff
@@ -905,21 +905,25 @@ class RandomSTE(torch.autograd.Function):
...
     generator = None
+    random_logits = None
 
     @staticmethod
     def forward(ctx, logits):
         """
         Forward pass returns random logits with rank-specific seed.
         """
+        if RandomSTE.random_logits is not None:
+            return RandomSTE.random_logits
```
Comment (Victarry) [inline megatron/core/transformer/moe/moe_utils.py]: I'm a little bit concerned of that always using the same random logits will introduction bias to the "balanced routing" if the first sampled logits are not very balanced.

Have you conducted experiments to show the perf diff or can we add a condition to the cuda graph case?

Comment (yaox12) [reply megatron/core/transformer/moe/moe_utils.py]: Added a condition for cuda graphs.

**Review by hxbai — APPROVED** [review]

**Comment by yaox12** [comment]
@Victarry @NVIDIA/core-nemo Can I get your approval?
